### PR TITLE
182 api endpoints for saas

### LINKED ISSUE #182: [SaaS Connectors] API endpoints for SaaS connector configuration
# Purpose
Modify or add API endpoints to fulfill the following:

1. Create a ConnectionConfig of type saas
2. Configure the connection with a saas_config (yaml)
3. Upload a corresponding dataset yaml for the saas connector
4. Run a privacy request via API for the configured SaaS Connector

### PR DESCRIPTION
**Pull request by galvana**
# Purpose
To augment the existing API endpoints to support SaaS config administration

# Changes
- Added `/validate_saas_config` and `/saas_config` endpoints to validate and add/update/remove SaaS configs to connections
- Added Postman collection for SaaS connectors

Validation checks enforce the following behavior for creation
1. create a connection config
2. add a SaaS config (to validate secrets and to provide references/identity information for the dataset)
3. add secrets (only if a SaaS config is available)
4. add dataset (only if a SaaS config is available, validates that the dataset doesn't contain references/identity information to prevent conflicts with SaaS config)

And the following behavior for deletion
1. delete a SaaS config (after deleting the dataset and secrets)

# Checklist

- [ ] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md))
- [ ] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Closes #182

### CODE REVIEW

**Comment by pattisdr** [comment]
Starting review!

File: src/fidesops/api/v1/endpoints/connection_endpoints.py [hunk 1]
```diff
@@ -177,12 +182,21 @@ def validate_secrets(
...
         connection_secrets = schema.parse_obj(request_body)
     except ValidationError as e:
-        raise HTTPException(
-            status_code=422,
-            detail=e.errors(),
-        )
+        raise HTTPException(status_code=422, detail=e.errors())
+    except SaaSConfigNotFoundException as e:
+        raise HTTPException(status_code=422, detail=e.message)
     return connection_secrets
```
Comment (pattisdr) [inline src/fidesops/api/v1/endpoints/connection_endpoints.py:200]: Question about the mailchimp config - it seems like secrets validation returns "succeeded" regardless of username and api_key.  Is validation happening properly there, or is this something specific related to the test account?

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py [hunk 2]
```diff
@@ -143,9 +148,19 @@ def patch_datasets(
             "dataset": dataset.dict(),
         }
         try:
+            if connection_config.connection_type == ConnectionType.saas:
+                __validate_saas_dataset(connection_config, dataset)
```
Comment (pattisdr) [inline src/fidesops/api/v1/endpoints/dataset_endpoints.py:152]: Another question about dataset validation for saas connection configs - you can technically add as many datasets as you want to a connection config.  Should there be an additional restriction that you can only add one if it is a saas connection config? Especially since they come in pairs.

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py [hunk 3]
```diff
@@ -161,6 +176,26 @@ def patch_datasets(
...
+    else:
+        for collection in dataset.collections:
+            for field in collection.fields:
+                graph_field = to_graph_field(field)
+                if graph_field.references or graph_field.identity:
+                    raise ValidationError(
+                        "A dataset for a ConnectionConfig type of saas is not "
+                        "allowed to have references or identities. Please add "
+                        "them to the SaaS config."
+                    )
```
Comment (pattisdr) [inline src/fidesops/api/v1/endpoints/dataset_endpoints.py:196]: Nice validation here - related, a leftover comment from your other PR - `SaaSConfig.get_graph` separately checks if there are references defined and if there is an identity defined, and if either case happens, a separate ScalarField is added.  https://github.com/ethyca/fidesops/pull/179#discussion_r807271127

1) I was wondering if someone could define a single field as both an identity, and a referenced field. If they made that mistake on the saas config, they would have two ScalarFields created.

2) Are we handling object and array fields with SaasConfigs yet? It seems like SaasConfig.get_graph can only create ScalarFields.

**Review by pattisdr — CHANGES_REQUESTED** [review]
First pass complete @galvana!  This was well-organized; I appreciate how the code adheres to some of the already-established patterns, and how you fleshed out some interesting edge cases in API unit tests. No big reorganization requested, most of my comments are around validation.

File: src/fidesops/api/v1/endpoints/connection_endpoints.py [hunk 4]
```diff
@@ -177,12 +182,21 @@ def validate_secrets(
     schema = get_connection_secrets_validator(connection_type.value, saas_config)
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/connection_endpoints.py:182]: That's exactly it, I missed that during the refactor.

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py [hunk 5]
```diff
@@ -161,6 +176,26 @@ def patch_datasets(
...
 
+def __validate_saas_dataset(
+    connection_config: ConnectionConfig, dataset: FidesopsDataset
+) -> None:
+    if connection_config.saas_config is None:
+        raise SaaSConfigNotFoundException(
+            f"Connection config '{connection_config.key}' must have a "
+            "SaaS config before adding a dataset"
+        )
+    else:
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/dataset_endpoints.py:187]: Fixed

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    SAAS_CONFIG_DELETE,
+    SAAS_CONFIG_READ,
+)
+
+from fidesops.api.v1.urn_registry import (
+    SAAS_CONFIG,
+    SAAS_CONFIG_VALIDATE,
+    V1_URL_PREFIX,
+)
+from fidesops.core.config import SecuritySettings
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:26]: Removed

File: src/fidesops/api/v1/endpoints/connection_endpoints.py [hunk 7]
```diff
@@ -177,12 +182,21 @@ def validate_secrets(
...
+        connection_type = connection_config.connection_type
+        saas_config = connection_config.get_saas_config()
+        if connection_type == ConnectionType.saas and saas_config is None:
+            raise SaaSConfigNotFoundException(
+                f"A SaaS config to validate the secrets is unavailable for this connection config, please add one via {SAAS_CONFIG}"
+            )
+        schema = get_connection_secrets_validator(connection_type.value, saas_config)
+        logger.info(
+            f"Validating secrets on connection config with key '{connection_config.key}'"
+        )
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/connection_endpoints.py:194]: Gotta be more careful with these refactors, thanks! I'm only going to log once after we get the validation schema and before we call `parse_obj`

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    saas_config: SaaSConfig,
+    db: Session = Depends(deps.get_db),
+    connection_config: ConnectionConfig = Depends(_get_connection_config),
+) -> ConnectionConfig:
+    """
+    Given a SaaS config element, update the corresponding ConnectionConfig object
+    or report failure
+    """
+    data = {"key": connection_config.key, "saas_config": saas_config.dict()}
+    logger.info(f"Starting upsert for SaaS config '{saas_config.fides_key}'")
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:81]: ```logger.info(f"Updating SaaS config '{saas_config.fides_key}' on connection config '{connection_config.key}'")```

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    return ConnectionConfig.create_or_update(db, data=data)
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:82]: Thanks, I'll use this syntax for updating fields moving forward.

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    dependencies=[Security(verify_oauth_client, scopes=[SAAS_CONFIG_READ])],
+    response_model=SaaSConfig,
+)
+def get_saas_config(
+    connection_config: ConnectionConfig = Depends(_get_connection_config),
+) -> SaaSConfig:
+    """Returns the SaaS config for the given connection config."""
+
+    logger.info(f"Finding SaaS config for connection '{connection_config.key}'")
+    saas_config = connection_config.saas_config
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:96]: I created the `_get_saas_connnection_config` function to perform an early validation and added it to the saas_config endpoints.

```This action is only applicable to connection configs of connection type 'saas'```

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+)
+def patch_saas_config(
+    saas_config: SaaSConfig,
+    db: Session = Depends(deps.get_db),
+    connection_config: ConnectionConfig = Depends(_get_connection_config),
+) -> ConnectionConfig:
+    """
+    Given a SaaS config element, update the corresponding ConnectionConfig object
+    or report failure
+    """
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:79]: Used the `_get_saas_connection_config` function to perform an early check.

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            msg=None,
+        ),
+    )
+
+
+@router.patch(
+    SAAS_CONFIG,
+    dependencies=[Security(verify_oauth_client, scopes=[SAAS_CONFIG_CREATE_OR_UPDATE])],
+    status_code=200,
+    response_model=ConnectionConfigurationResponse,
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:69]: Updated this endpoint to return a SaaSConfig, there are no secrets there.

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    Checks that:
+    - all required fields are present, all field values are valid types
+    """
+
+    logger.info(f"Validation successful for SaaS config '{saas_config.fides_key}'!")
+    return ValidateSaaSConfigResponse(
+        saas_config=saas_config,
+        validation_details=SaaSConfigValidationDetails(
+            msg=None,
+        ),
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:61]: I was following the pattern we've established for `ValidateDatasetResponse` even though we don't have any validation happening right now. But if we wanna follow the YAGNI principle I can have it just return the SaaS config.

[the remark this answers is not in the document]
Comment (pattisdr) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:61]: hmm, okay, could you add something to the msg then, stating how the validation was successful?

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    saas_config: SaaSConfig,
+) -> ValidateSaaSConfigResponse:
+    """
+    Run validations against a SaaS config without attempting to save it to the database.
+
+    Checks that:
+    - all required fields are present, all field values are valid types
+    """
+
+    logger.info(f"Validation successful for SaaS config '{saas_config.fides_key}'!")
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:56]: Using inside voice for the logger message and updated the docstring

```
    Uses the SaaSConfig Pydantic model to validate the SaaS config
    without attempting to save it to the database.
```

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+        )
+
+    fides_key = saas_config["fides_key"]
+    dataset = DatasetConfig.filter(
+        db=db,
+        conditions=(
+            (DatasetConfig.connection_config_id == connection_config.id)
+            & (DatasetConfig.fides_key == fides_key)
+        ),
+    ).first()
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:132]: Added a check to make sure the `fides_key`s match.

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+    warnings = []
+    if dataset:
+        warnings.append(
+            f"Must delete the dataset with fides_key '{fides_key}' before deleting this SaaS config."
+        )
+    if connection_config.secrets:
+        warnings.append(
+            "Must clear the secrets from this connection config before deleting the SaaS config."
+        )
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:142]:
```
# The secrets must be cleared since the SaaS config is used for validation and the secrets
# might not pass validation once a new SaaS config is added.
```

File: docs/fidesops/docs/postman/Fidesops.postman_collection.json [hunk 17]
```diff
@@ -2237,6 +2237,191 @@
...
+									"value": "{{client_token}}",
+									"type": "string"
+								}
+							]
+						},
+						"method": "PUT",
+						"header": [],
+						"body": {
+							"mode": "raw",
+							"raw": "{\n    \"domain\": \"\",\n    \"username\": \"\",\n    \"api_key\": \"\"\n}",
```
[the remark this answers is not in the document]
Comment (galvana) [reply docs/fidesops/docs/postman/Fidesops.postman_collection.json:2336]: Modified the body to use variables.

File: src/fidesops/api/v1/endpoints/connection_endpoints.py [hunk 18]
```diff
@@ -177,12 +182,21 @@ def validate_secrets(
     schema = get_connection_secrets_validator(connection_type.value, saas_config)
 
     try:
+        connection_type = connection_config.connection_type
+        saas_config = connection_config.get_saas_config()
+        if connection_type == ConnectionType.saas and saas_config is None:
+            raise SaaSConfigNotFoundException(
+                f"A SaaS config to validate the secrets is unavailable for this connection config, please add one via {SAAS_CONFIG}"
+            )
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/connection_endpoints.py:190]: Thanks for the suggestion, that's a lot cleaner.

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py (hunk 2 above)
Comment (galvana) [reply src/fidesops/api/v1/endpoints/dataset_endpoints.py:152]: Enforcing this by checking for matching fides_keys between the dataset and the SaaS config.

There is still the scenario where the user can submit multiple valid datasets with the same fides_key but they would override each other as we iterated through the list. Not sure how to notify the user of this while still conforming to the `BulkPutDataset` response format. Would a `logger.warn` be sufficient?

File: docs/fidesops/docs/postman/Fidesops.postman_collection.json [hunk 19]
```diff
@@ -2237,6 +2237,191 @@
...
+							"path": [
+								"connection",
+								""
+							]
+						}
+					},
+					"response": []
+				},
+				{
+					"name": "Create/Update SaaS Config",
```
[the remark this answers is not in the document]
Comment (galvana) [reply docs/fidesops/docs/postman/Fidesops.postman_collection.json:2282]: Added!

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+
+    if warnings:
+        raise HTTPException(status_code=HTTP_400_BAD_REQUEST, detail=" ".join(warnings))
+
+    logger.info(f"Deleting SaaS config for connection '{connection_config.key}'")
+    ConnectionConfig.create_or_update(
+        db, data={"key": connection_config.key, "saas_config": None}
+    )
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:150]: Updated!

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+
+
+@router.delete(
+    SAAS_CONFIG,
+    dependencies=[Security(verify_oauth_client, scopes=[SAAS_CONFIG_DELETE])],
+    status_code=204,
+)
+def delete_saas_config(
+    db: Session = Depends(deps.get_db),
+    connection_config: ConnectionConfig = Depends(_get_connection_config),
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:112]: Added.

File: src/fidesops/api/v1/endpoints/connection_endpoints.py [hunk 22]
```diff
@@ -177,12 +182,21 @@ def validate_secrets(
...
+        schema = get_connection_secrets_validator(connection_type.value, saas_config)
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/connection_endpoints.py:191]: Added `test_connection_secrets_saas.py`

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:61]: Added a `Validation successful` message.

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py (hunk 3 above)
Comment (galvana) [reply src/fidesops/api/v1/endpoints/dataset_endpoints.py:196]: Added a validation to ConnectorParam which throws a `ValueError` if both `references` and `identities` are defined for the same param.

Comment (galvana) [reply src/fidesops/api/v1/endpoints/dataset_endpoints.py:196]: As for the object and array fields, we can keep using scalar fields for the SaaS config since all we need is a place to hold references and identities to be able to add them to the dataset.

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py (hunk 2 above)
Comment (pattisdr) [reply src/fidesops/api/v1/endpoints/dataset_endpoints.py:152]: that's a good idea

File: src/fidesops/api/v1/endpoints/connection_endpoints.py (hunk 1 above)
Comment (galvana) [reply src/fidesops/api/v1/endpoints/connection_endpoints.py:200]: I wasn't handing non HTTP status codes correctly, the connection test fails now when providing incorrect credentials.

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py [hunk 23]
```diff
@@ -161,6 +195,33 @@ def patch_datasets(
     )
 
 
+def __validate_saas_dataset(
```
Comment (seanpreston) [inline src/fidesops/api/v1/endpoints/dataset_endpoints.py:198]: Private methods should only append one `_`

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+
+    logger.info(f"Finding SaaS config for connection '{connection_config.key}'")
+    saas_config = connection_config.saas_config
+    if not saas_config:
+        raise HTTPException(
+            status_code=HTTP_404_NOT_FOUND,
+            detail=f"No SaaS config found for connection '{connection_config.key}'",
+        )
+
+    fides_key = saas_config["fides_key"]
```
Comment (seanpreston) [inline src/fidesops/api/v1/endpoints/saas_config_endpoints.py:142]: I know we're using schemas to validate this data on entry, but since this is a retrieval it could be susceptible to dirty data. It's worth considering `.get("fides_key", <default>)` here

**Comment by pattisdr** [comment]
Starting second round of review @galvana!

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py [hunk 25]
```diff
@@ -87,9 +92,19 @@ def validate_dataset(
...
+        graph = convert_dataset_to_graph(dataset, connection_config.key)
+
+        # Datasets for SaaS connections need to be merged with a SaaS config to
+        # be able to generate a valid traversal
+        if connection_config.connection_type == ConnectionType.saas:
+            __validate_saas_dataset(connection_config, dataset)
+            graph = merge_datasets(
+                graph,
+                connection_config.get_saas_config().get_graph(),
+            )
```
Comment (pattisdr) [inline src/fidesops/api/v1/endpoints/dataset_endpoints.py:104]: For future consideration, I still think it would be nice to see a representation of the merged dataset somewhere in an API response.  If there are issues, being able to see how we merged the dataset might be useful.  However, I understand that the merged dataset is of type Dataset, not FidesopsDataset, making it difficult to return the merge here, so this could be a follow-up.

This is a positive step forward though, merging the Datasets before building the traversal for saas configs.

**Review by pattisdr — COMMENTED** [review]
Nice changes here @galvana, lots of improved validation to help users make sure their saas configs and datasets are working as intended! just a couple of questions from me, it's about ready to go on my end

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py (hunk 23 above)
Comment (galvana) [reply src/fidesops/api/v1/endpoints/dataset_endpoints.py:198]: Updating to use only one `_`.

File: src/fidesops/service/connectors/saas_connector.py [hunk 26]
```diff
@@ -76,15 +74,20 @@ def query_config(self, node: TraversalNode) -> SaaSQueryConfig:
...
+            response = self.client().get(prepared_request)
         except Exception:
             raise ConnectionException(
-                f"Operational Error connecting to {self.configuration.key}."
+                f"Operational Error connecting to '{self.configuration.key}'."
             )
 
+        if not response.ok:
+            raise ClientUnsuccessfulException(status_code=response.status_code)
+
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/service/connectors/saas_connector.py:90]: Nice catch, I looked this up in the Mailchimp API documentation and apparently we can authenticate using one of two ways.

1. Bearer auth using the api_key
2. Basic auth with "anystring" as the username, and the api_key as the password

I'll keep the basic auth for now but I think this would be cleaner if we used bearer auth in the future to prevent confusion.

File: src/fidesops/api/v1/endpoints/dataset_endpoints.py [hunk 27]
```diff
@@ -136,21 +151,40 @@ def patch_datasets(
     created_or_updated: List[FidesopsDataset] = []
     failed: List[BulkUpdateFailed] = []
     logger.info(f"Starting bulk upsert for {len(datasets)} datasets")
+
+    # warn if there are duplicate fides_keys within the datasets
+    # valid datasets with the same fides_key will override each other
+    key_list = [dataset.fides_key for dataset in datasets]
+    if len(key_list) != len(set(key_list)):
+        logger.warning(
+            "Datasets with duplicate fides_keys detected, may result in unintented behavior."
```
[the remark this answers is not in the document]
Comment (galvana) [reply src/fidesops/api/v1/endpoints/dataset_endpoints.py:160]: Fixed and added spell checker extension to prevent these from getting through again.

File: src/fidesops/api/v1/endpoints/saas_config_endpoints.py (hunk 24 above)
Comment (galvana) [reply src/fidesops/api/v1/endpoints/saas_config_endpoints.py:142]: Using `.get("fides_key")` and throwing an early exception if the value is None.

**Comment by pattisdr** [comment]
Before release, I think guides for these endpoints would be important - also guides on how to create the saas configs and saas datasets.

File: src/fidesops/service/connectors/saas_connector.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (pattisdr) [reply src/fidesops/service/connectors/saas_connector.py]: cool thanks for digging into that

File: src/fidesops/service/connectors/saas_connector.py [hunk 28]
```diff
@@ -74,20 +84,9 @@ def query_config(self, node: TraversalNode) -> SaaSQueryConfig:
...
-        except Exception:
-            raise ConnectionException(
-                f"Operational Error connecting to '{self.configuration.key}'."
-            )
-
-        if not response.ok:
-            raise ClientUnsuccessfulException(status_code=response.status_code)
-
+        prepared_request: SaaSRequestParams = test_request_path, {}, {}
+        self.client().get(prepared_request)
```
Comment (pattisdr) [inline src/fidesops/service/connectors/saas_connector.py:89]: Nice sharing this code 👍

**Review by pattisdr — APPROVED** [review]
good work here @galvana!

A lingering concern is that we may still be special-casing SaaS Configs too frequently; SaaS Configs are continuing to diverge from how our other Connection Configs work, and it may not always be necessary, it's just out of design (I'm talking about things like saas configs can only have references point in one direction)  That being said, I don't think that's a reason to hold this up; this is a nice set of API endpoints that let us create the needed config to make requests to third-party endpoints. We can continue to refine later.
